common/clock/event_time_source.go

```diff
@@ -0,0 +1,159 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package clock
+
+import (
+	"sync"
+	"time"
+)
+
+type (
+	// EventTimeSource is a timex.TimeSource. Unlike other fake timeSource implementations, the methods of this
+	// timeSource are synchronous, so when you call Advance or Update, all triggered timers from AfterFunc will fire
+	// before the method returns, in the same goroutine.
+	EventTimeSource struct {
+		mu     sync.RWMutex
+		now    time.Time
+		timers []*fakeTimer
+	}
+
+	// fakeTimer is a fake implementation of [timex.Timer].
+	fakeTimer struct {
+		// need a link to the parent timeSource for synchronization
+		timeSource *EventTimeSource
+		// deadline for when the timer should fire
+		deadline time.Time
+		// callback to call when the timer fires
+		callback func()
+		// done is true if the timer has fired or been stopped
+		done bool
+		// index of the timer in the parent timeSource
+		index int
+	}
+)
+
+// NewEventTimeSource returns a EventTimeSource with the current time set to Unix zero: 1970-01-01 00:00:00 +0000 UTC.
+func NewEventTimeSource() *EventTimeSource {
+	return &EventTimeSource{
+		now: time.Unix(0, 0),
+	}
+}
+
+// Now return the current time.
+func (ts *EventTimeSource) Now() time.Time {
+	ts.mu.RLock()
+	defer ts.mu.RUnlock()
+
+	return ts.now
+}
+
+// AfterFunc return a timer that will fire after the specified duration. It is important to note that the timeSource is
+// locked while the callback is called. This means that you must be cautious about calling any methods on the timeSource
+// from within the callback. Doing so will probably result in a deadlock. To avoid this, you may want to wrap all such
+// calls in a goroutine.
+func (ts *EventTimeSource) AfterFunc(d time.Duration, f func()) Timer {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	t := &fakeTimer{timeSource: ts, deadline: ts.now.Add(d), callback: f}
+	t.index = len(ts.timers)
+	ts.timers = append(ts.timers, t)
+
+	return t
+}
+
+// Update the fake current time. It returns the timeSource so that you can chain calls like this:
+// timeSource := NewEventTimeSource().Update(time.Now())
+func (ts *EventTimeSource) Update(now time.Time) *EventTimeSource {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	ts.now = now
+	ts.fireTimers()
+	return ts
+}
+
+// Advance the timer by the specified duration.
+func (ts *EventTimeSource) Advance(d time.Duration) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	ts.now = ts.now.Add(d)
```

Comment:
Added. I made it so we just fire the timer immediately. The real time package will set the timer to fire now. I make it so that the timer fires before AfterFunc even returns.